extensions/amp-web-push/0.1/amp-web-push-config.js

```diff
@@ -0,0 +1,204 @@
+  SERVICE_WORKER_URL: 'serviceWorkerUrl',
+};
+
+/**
+ * @fileoverview
+ * A <amp-web-push-config> element that exposes attributes for publishers to
+ * configure the web push service.
+ *
+ * On buildCallback(), the element starts the web push service.
+ *
+ * Only a single <amp-web-push-config> is allowed in the document, and it must
+ * have the ID "amp-web-push". Subscribe and unsubscribe actions dispatched from
+ * various <amp-web-push-widgets> are all processed by <amp-web-push-config>
+ * which then forwards the event to the web push service.
+ */
+export class WebPushConfig extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.NODISPLAY;
+  }
+
+  validate() {
+    this.ensureSpecificElementId_();
+    this.ensureUniqueElement_();
+
+    const config = {
+      helperIframeUrl: null,
+      permissionDialogUrl: null,
+      serviceWorkerUrl: null,
+    };
+
+    for (const attribute in WebPushConfigAttributes) {
+      const value = WebPushConfigAttributes[attribute];
+      user().assert(this.element.getAttribute(value),
+          `The ${value} attribute is required for <${CONFIG_TAG}>`);
+      config[value] = this.element.getAttribute(value);
+    }
+
+    if (!this.isValidHelperOrPermissionDialogUrl_(config['helperIframeUrl'])) {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'helperIframeUrl attribute. It should begin with ' +
+        'the https:// protocol and point to the provided lightweight ' +
+        'template page provided for AMP messaging.');
+    }
+
+    if (!this.isValidHelperOrPermissionDialogUrl_(
+        config['permissionDialogUrl'])) {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'permissionDialogUrl attribute. It should begin with ' +
+        'the https:// protocol and point to the provided template page ' +
+        'for showing the permission prompt.');
+    }
+
+    if (parseUrl(config['serviceWorkerUrl']).protocol !== 'https:') {
+      throw user().createError(`<${CONFIG_TAG}> must have a valid ` +
+        'serviceWorkerUrl attribute. It should begin with the ' +
+        'https:// protocol and point to the service worker JavaScript file ' +
+        'to be installed.');
+    }
+
+    if (parseUrl(config['serviceWorkerUrl']).origin !==
+          parseUrl(config['permissionDialogUrl']).origin ||
+        parseUrl(config['permissionDialogUrl']).origin !==
+        parseUrl(config['helperIframeUrl']).origin) {
+      throw user().createError(`<${CONFIG_TAG}> URL attributes ` +
+        'serviceWorkerUrl, permissionDialogUrl, and ' +
+        'helperIframeUrl must all share the same origin.');
+    }
+  }
+
+  /**
+   * Ensures this element is defined with TAG id.
+   *
+   * @private
+   */
+  ensureSpecificElementId_() {
+    if (this.element.getAttribute('id') !== TAG) {
```

Comment:
I think this is necessary. This is supposed to check that any created <amp-web-push-config> element has only one allowed ID "amp-web-push", so that tap actions can be routed to this config element (if we are keeping the `<amp-web-push-config>` element).